newIDE/app/public/external/yarn/yarn-main.js

```diff
@@ -0,0 +1,77 @@
+import { createPathEditorHeader, fileExists } from '../utils/path-editor.js';
+
+const electron = require('electron');
+const electronWindow = electron.remote.getCurrentWindow();
+const ipcRenderer = electron.ipcRenderer;
+const fs = require('fs');
+const remote = electron.remote;
+
+let yarn = null;
+let receivedData;
+
+const saveAndClose = pathEditor => {
+  const savePath = pathEditor.state.fullPath;
+  yarn.data.saveTo(savePath, yarn.data.getSaveData('json'), () => {
+    ipcRenderer.send('yarn-changes-saved', savePath, receivedData);
+    remote.getCurrentWindow().close();
+  });
+};
+
+const closeWindow = () => {
+  remote.getCurrentWindow().close();
+};
+
+const editorFrameEl = document.getElementById('yarn-frame');
+window.addEventListener('yarnReady', e => {
+  yarn = e;
+  yarn.app.fs = fs;
+  yarn.app.electron = electron;
+  ipcRenderer.send('yarn-ready');
+});
+editorFrameEl.src = 'yarn-editor/index.html';
+
+// Called to load yarn data. Should be called after the window is fully loaded.
+
+ipcRenderer.on('yarn-open', (event, receivedData) => {
+  //Make a header
+  const pathEditorHeaderDiv = document.getElementById('path-editor-header');
+  const pathEditorHeader = createPathEditorHeader({
+    parentElement: pathEditorHeaderDiv,
+    editorContentDocument: document,
+    onSaveToGd: saveAndClose,
+    onCancelChanges: closeWindow,
+    projectPath: receivedData.projectPath,
+    initialResourcePath: receivedData.resourcePath,
+    extension: '.json',
+  });
+  //inject custom save+close button
+  const saveToGdButton = yarn.document
+    .getElementsByClassName('menu')[0]
+    .cloneNode(true);
+  saveToGdButton.onclick = () => saveAndClose(pathEditorHeader);
+  yarn.document
+    .getElementsByClassName('app-menu')[0]
+    .appendChild(saveToGdButton);
+  saveToGdButton.childNodes[0].firstChild.data = 'Save & close';
+
+  // process the json file,if it exists
+  if (fileExists(receivedData.resourcePath)) {
+    receivedData.externalEditorData = fs
+      .readFileSync(receivedData.resourcePath, 'utf8')
+      .toString();
+
+    yarn.data.loadData(receivedData.externalEditorData, 'json', true);
+    electronWindow.setTitle(
+      'GDevelop Dialogue Tree Editor - ' + receivedData.resourcePath
+    );
+
+    pathEditorHeader.toggle();
+  } else {
+    // If GD has sent no path, we need to set one for yarn
+    receivedData.resourcePath = receivedData.projectPath + '/NewFile.json';
+  }
+
+  receivedData = receivedData;
```

Comment:
This is surely broken because `receivedData = receivedData` is not changing the `receivedData` outside of the function scope. The variable outside the function is said to be "shadowed" by the parameter.

`receivedData` is a good name for the parameter. But for the variable outside the function, why is that called "received"? Is it really it purpose? What does this thing contains? Try to answer to these questions to find a proper name :) 

For finding a good name, inspect what the variable contains and what it is used for. I can see it's passed to the event "yarn-changes-saved".  Then when I look at where it's used, it seems to be passed and then, in `ipcRenderer.on('yarn-changes-saved', [...]` to be... never used ^^

In this case, if you don't need it, remove it and pass `null` - if that make sense in the code :)

Reply:
(I understand this is a detail, but I want to be sure that the code handling the external editor is as simple as possible.)

Reply:
> > I noticed that your zipped file contains a new folder called _MACOSX
> > Is that intended? My code doesn't use it at all and it requires adding a .gitignore line
> 
> You can re-create a proper zip without it - it's automatically added by macOS Very disturbing - it's hidden by the macOS Finder.
> 
> > it's still giving a warning for some reason, but does download and unzip it
> 
> Weird, are you using the exact same file as the downloaded one? Any change in any bit in the file will totally change the checksum :) I'll update it myself if needed.

The _macosx folder is in the zip file you uploaded. You might have to reupload it :)

Reply:
Yeah if you give me a proper updated zip (wihtout the macos folder) I'll update the one hosted.

Reply:
removed the receivedData variable and passing null. You are right - it is redundant in this case